### PR TITLE
remove __unmatch [run_process_replay]

### PR DESCRIPTION
**Pull request by DKormann**

### CODE REVIEW

**Comment by Qazalin** [comment]
did you benchmark this?

**Comment by DKormann** [comment]
> did you benchmark this?

yeah it is a little slower

**Comment by geohot** [comment]
I liked the simpler __unmatch function but that list construction is awful.

**Comment by geohot** [comment]
Is redefining the function inside the function slower? Also, be careful with your spacing.

**Comment by DKormann** [comment]
its not slower for me:
➜  tinygrad git:(simplify_match) ✗ PROFILE=0 python3 test/external/external_benchmark_schedule.py
***** model lower in 1103.99 ms
***** model lower in 1101.71 ms
***** model lower in 1094.21 ms
with inside function: 
***** model lower in 1104.00 ms
***** model lower in 1090.17 ms
***** model lower in 1095.75 ms

File: tinygrad/codegen/uops.py [hunk 1]
```diff
@@ -105,19 +105,11 @@ def compile(u: UOp, name:Optional[str]=None) -> UPat:
...
-  elif m2 != m1: return True
-  return False
-
 def _match(uop:UOp, pat:UPat, store:Dict[str, UOp]) -> bool:
   if pat.name in store and store[pat.name] is not uop: return False
   if pat.name is not None: store[pat.name] = uop
-  if pat.arg is not None and __unmatch(pat.arg, uop.arg): return False
-  if pat.dtype is not None and uop.dtype is not None and __unmatch(pat.dtype, uop.dtype): return False
-  if pat.op is not None and __unmatch(pat.op, uop.op): return False
+  def unmatch(p, u): return p is not None and (u not in p if isinstance(p, set) else u != p)
```
[the remark this answers is not in the document]
Comment (DKormann) [reply tinygrad/codegen/uops.py]: sure

**Comment by Qazalin** [comment]
I'd add type annotations back

File: tinygrad/codegen/uops.py [hunk 2]
```diff
@@ -106,18 +106,12 @@ def compile(u: UOp, name:Optional[str]=None) -> UPat:
     return UPat(u.op, u.arg, (list if u.commutative() else tuple)([UPat.compile(src) for src in u.src]) if u.src != () else None, name, u.dtype)
 
 T = TypeVar("T")
-def __unmatch(m1:Union[T, Set[T]], m2:T) -> bool:
-  if isinstance(m1, set):
-    if m2 not in m1: return True
-  elif m2 != m1: return True
-  return False
+def __unmatch(p:Union[T, Set[Optional[T]]], u:T): return p is not None and (u not in p if isinstance(p, Set) else u != p)
```
Comment (DKormann) [inline tinygrad/codegen/uops.py:109]: can only get mypy with Set(Optional[T]) not sure why

**Comment by Qazalin** [comment]
Set[Optional[T]] is clearly wrong,
you should do this:
```
def __unmatch(p:Union[T, Set[T], None], u:Optional[T]) -> bool
```

**Comment by Qazalin** [comment]
wait this is much slower than what you claim

PROFILE=0 python3 test/external/external_benchmark_schedule.py
master:
```
***** model forward in  54.08 ms
***** model schedule in  17.19 ms
***** model lower in 2164.34 ms
```

this branch:
```
***** model forward in  53.24 ms
***** model schedule in  17.13 ms
***** model lower in 2516.40 ms
```

tested on tinybox green

**Comment by DKormann** [comment]
that was in regard to outside vs inside __unmatch funciton not comparing to master

**Comment by DKormann** [comment]
did run my tests on m2 pro

**Comment by DKormann** [comment]
don't think can speed this up more its slower than expected
feel free to close if not worth it

**Comment by DKormann** [comment]
around 1090ms -> 1285 on m2

**Comment by Qazalin** [comment]
I think those lines existed for saving performance,
I wouldn't merge a 15% perf regression just to have 6 less lines,

**Pull request closed without merging**